src/_P037_MQTTImport.ino

```diff
@@ -138,66 +223,320 @@ boolean Plugin_037(byte function, struct EventStruct *event, String& string)
+              ++iter;
+            } while (processData && iter != doc.end());
+            iter = doc.begin();
+          }
+#endif // ifndef P037_FILTER_PER_TOPIC
+#endif // P037_JSON_SUPPORT
+        }
+        if (matchedTopic && P037_data->hasFilters()) { // Single log statement
+          if (loglevelActiveFor(LOG_LEVEL_INFO)) {
+            log = F("IMPT : MQTT 037 filter result: ");
+            log += processData ? F("true") : F("false");
+            addLog(LOG_LEVEL_INFO, log);
+          }
+        }
+#endif // P037_FILTER_SUPPORT
 
-            // Generate event for rules processing - proposed by TridentTD
-
-            if (Settings.UseRules)
-            {
-              String RuleEvent;
-              RuleEvent += getTaskDeviceName(event->TaskIndex);
-              RuleEvent += '#';
-              RuleEvent += ExtraTaskSettings.TaskDeviceValueNames[x];
-              RuleEvent += '=';
-              RuleEvent += floatPayload;
-              eventQueue.add(RuleEvent);
+        // Get the Topic and see if it matches any of the subscriptions
+        for (byte x = 0; x < VARS_PER_TASK && processData; x++)
+        {
+          String subscriptionTopic = P037_data->StoredSettings.deviceTemplate[x];
+          subscriptionTopic.trim();
+          if (subscriptionTopic.length() == 0) continue;							// skip blank subscriptions
+
+          // Now check if the incoming topic matches one of our subscriptions
+          parseSystemVariables(subscriptionTopic, false);
+          if (MQTTCheckSubscription_037(event->String1, subscriptionTopic)) {
+#ifdef P037_JSON_SUPPORT
+#ifdef P037_FILTER_PER_TOPIC
+            // json filter check
+            bool passFilter = true;
+            if (checkJson && P037_data->hasFilters()) { // See if we pass the filters for all json attributes
+              do {
+                key = iter->key().c_str();
+                Payload = iter->value().as<String>();
+#ifdef P037_MAPPING_SUPPORT
+                if (P037_APPLY_MAPPINGS) {
+                  Payload = P037_data->mapValue(Payload, key);
+                }
+#endif
+                passFilter = P037_data->checkFilters(key, Payload, x + 1); // Will return true unless key matches *and* Payload doesn't
+                ++iter;
+              } while (passFilter && iter != doc.end());
+              iter = doc.begin();
             }
+            if (passFilter) // Watch it, no curly braces, so only 1 statement processed: (do {...) !
+#endif // P037_FILTER_PER_TOPIC
+#endif // P037_JSON_SUPPORT
+            do {
+#ifdef P037_JSON_SUPPORT
+              if (checkJson && iter != doc.end()) {
+                String jsonAttribute = P037_data->StoredSettings.jsonAttributes[x];
+                jsonAttribute.replace(';', ',');
+                String jsonIndex = parseString(jsonAttribute, 2);
+                jsonAttribute = parseString(jsonAttribute, 1);
+                jsonAttribute.trim();
+                if (jsonAttribute.length() > 0) {
```

Comment:
Upto here, the parsing will result in the same attribute and index every single time, until settings are changed.

Reply:
Moved to initialization.